Bluetooth: Classic: Add PBAP (Phone Book Access Profile) support

CC0918 opened the pull request:
Add Phone Book Access Profile (PBAP) implementation for Bluetooth Classic,
supporting both client (PCE - Phone Book Client Equipment) and server
(PSE - Phone Book Server Equipment) roles.

PBAP enables access to phone book objects stored on a remote device,
allowing retrieval of contact information in vCard format.

Key features:
- PCE (client) role: connect, disconnect, pull phonebook, pull vCard
  listing, pull vCard entry, set path, and abort operations
- PSE (server) role: RFCOMM and L2CAP server registration, handling
  client requests and sending responses
- Authentication support using MD5-based challenge-response mechanism
- Single Response Mode (SRM) support for improved throughput over L2CAP
- Application parameter handling for filtering and formatting requests
- Transport layer abstraction supporting both RFCOMM and L2CAP

The implementation includes:
- Core PBAP protocol handling (subsys/bluetooth/host/classic/pbap.c)
- Public API header (include/zephyr/bluetooth/classic/pbap.h)
- Shell commands for testing both PCE and PSE roles
- OBEX header extensions for authentication
- Kconfig options with mbedTLS MD5 dependency
- Test configuration updates

lylezhu2012 commented on include/zephyr/bluetooth/classic/pbap.h:
```diff
@@ -0,0 +1,636 @@
+	BT_PBAP_RSP_CODE_NOT_IMPL = 0xd1,
+	/* Service Unavailable */
+	BT_PBAP_RSP_CODE_UNAVAIL = 0xd3,
+};
```
Suggest mapping to OBEX response code.

lylezhu2012 commented on include/zephyr/bluetooth/classic/pbap.h:
```diff
@@ -0,0 +1,917 @@
+/** @brief PBAP response Code. */
+enum __packed bt_pbap_rsp_code {
+	/* Continue */
+	BT_PBAP_RSP_CODE_CONTINUE = BT_OBEX_RSP_CODE_CONTINUE,
```
Please change the comments from `/* ... */` to `/** ... */`

lylezhu2012 commented on include/zephyr/bluetooth/classic/pbap.h:
```diff
@@ -0,0 +1,917 @@
+	BT_PBAP_TRANSPORT_STATE_CONNECTED = 2,
+	/** @brief Transport is disconnecting */
+	BT_PBAP_TRANSPORT_STATE_DISCONNECTING = 3,
+};
```
Ditto.

lylezhu2012 commented on include/zephyr/bluetooth/classic/pbap.h:
```diff
@@ -0,0 +1,917 @@
+
+	/** @internal Transport state (atomic) */
+	atomic_t _transport_state;
+};
```
I think the structure should be moved before the deceleration of function `bt_pbap_pce_rfcomm_connect()`.

lylezhu2012 commented on subsys/bluetooth/host/classic/Kconfig:
```diff
@@ -707,6 +707,27 @@ config BT_BIP
+	help
+	  This option enables the PCE
+
+endif #BT_PBAP
```
```suggestion
config BT_PBAP_PSE
	bool "Bluetooth PBAP Profile Server [EXPERIMENTAL]"
	select BT_PBAP
	help
	  This option enables the PSE

config BT_PBAP_PCE
	bool "Bluetooth PBAP Profile Client [EXPERIMENTAL]"
	select BT_PBAP
	help
	  This option enables the PCE
```

CC0918 committed:
bluetooth: classic: obex: add length defines for digest tags

Add length constant definitions for OBEX digest-challenge and
digest-response tags to improve code readability and maintainability.

The following length defines are added:
- BT_OBEX_CHALLENGE_TAG_NONCE_LEN (16 bytes)
- BT_OBEX_CHALLENGE_TAG_OPTIONS_LEN (1 byte)
- BT_OBEX_RESPONSE_TAG_REQ_DIGEST_LEN (16 bytes)
- BT_OBEX_RESPONSE_TAG_NONCE_LEN (16 bytes)

Signed-off-by: Cheng Chang <cheng.chang@nxp.com>

lylezhu2012 commented on include/zephyr/bluetooth/classic/pbap.h:
```diff
@@ -0,0 +1,950 @@
+};
+
+/** @brief The PBAP PropertySelector bitmask. */
+enum bt_pbap_appl_param_property_mask {
```
My suggestion is to replace `BIT()` with `BIT64()`.

lylezhu2012 commented on subsys/bluetooth/host/classic/pbap.c:
```diff
@@ -0,0 +1,1425 @@
+	if (status != PSA_SUCCESS) {
+		LOG_WRN("Generate auth response failed %d", status);
+		return status;
+	}
```
I do not think the error code of `PSA` can be mixed with `POSIX`.

lylezhu2012 commented on subsys/bluetooth/host/classic/Kconfig:
```diff
@@ -720,6 +720,25 @@ config BT_BIP
+	select BT_PBAP
+	help
+	  This option enables the PSE
+config BT_PBAP_PCE
```
New empty line is missing.

lylezhu2012 commented on subsys/bluetooth/host/classic/pbap.c:
```diff
@@ -0,0 +1,1425 @@
+				LOG_ERR("Failed to disconnect %d", err);
+			}
+			return;
+		}
```
If the who is mismatched, the transport will be disconnected by OBEX. So the condition should not be reached here.

lylezhu2012 replied:
If the who is mismatched, the transport will be disconnected by OBEX. So the condition should not be reached here.

lylezhu2012 commented on subsys/bluetooth/host/classic/pbap.c:
```diff
@@ -0,0 +1,1425 @@
+
+static void pbap_pce_setpath(struct bt_obex_client *client, uint8_t rsp_code, struct net_buf *buf)
+{
+
```
Invalid empty line.

lylezhu2012 commented on subsys/bluetooth/host/classic/pbap.c:
```diff
@@ -0,0 +1,1425 @@
+#define PBAP_PWD_MAX_LENGTH 16U
+
+NET_BUF_POOL_FIXED_DEFINE(bt_pbap_pool, CONFIG_BT_MAX_CONN,
+			  BT_RFCOMM_BUF_SIZE(CONFIG_BT_GOEP_RFCOMM_MTU),
```
Shall we consider using a maximum value between `CONFIG_BT_GOEP_RFCOMM_MTU` and `CONFIG_BT_GOEP_L2CAP_MTU`.

lylezhu2012 commented on subsys/bluetooth/host/classic/pbap.c:
```diff
@@ -0,0 +1,1425 @@
+	int32_t status = PSA_SUCCESS;
+
+	status = bt_pbap_generate_auth_response(pwd, auth_chal_req, (uint8_t *)&result);
+	if (status == PSA_SUCCESS) {
```
```suggestion
	if (status == 0) {
```

lylezhu2012 commented on subsys/bluetooth/host/classic/pbap.c:
```diff
@@ -0,0 +1,1425 @@
+{
+	struct bt_pbap_pse *pbap_pse = CONTAINER_OF(server, struct bt_pbap_pse, _server);
+
+	if (pbap_pse != NULL && pbap_pse->cb != NULL && pbap_pse->cb->abort != NULL) {
```
```suggestion
	if (pbap_pse->cb != NULL && pbap_pse->cb->abort != NULL) {
```

lylezhu2012 commented on subsys/bluetooth/host/classic/pbap.c:
```diff
@@ -0,0 +1,1425 @@
+	if (err != 0) {
+		LOG_ERR("Failed to send get rsp %d", err);
+		return err;
+	}
```
I think there is checking missing. The precondition of `get` response is there is a pending `get` request. It should be checked before calling `bt_obex_get_rsp()`.

lylezhu2012 commented on subsys/bluetooth/host/classic/shell/pbap.c:
```diff
@@ -0,0 +1,2782 @@
+	}
+
+	return add_header_appl_param(BT_PBAP_APPL_PARAM_TAG_ID_SUPPORTED_FEATURES, sizeof(data),
+				     data, last);
```
Is there any connection between the features set here and the feature defined in SDP? If yes, I think all of them should be consistent.

lylezhu2012 commented on subsys/bluetooth/host/classic/shell/pbap.c:
```diff
@@ -0,0 +1,2782 @@
+		last = true;
+	}
+
+	return add_header_appl_param(BT_PBAP_APPL_PARAM_TAG_ID_SUPPORTED_FEATURES, sizeof(data),
```
Is there any connection between the features set here and the feature defined in SDP? If yes, I think all of them should be consistent.

CC0918 commented on include/zephyr/bluetooth/classic/pbap.h:
```diff
@@ -0,0 +1,950 @@
+ */
+#define PSE_ASSUMED_SUPPORT_FEATURE                                                                \
+	BT_PBAP_SUPPORTED_FEATURE_DOWNLOAD || BT_PBAP_SUPPORTED_FEATURE_BROWSING
+
```
The role is PSE, when it receives connect req without  pbap supported features parameter  from pce, pse should determind the featur of oce is 0x03.

CC0918 commented on include/zephyr/bluetooth/classic/pbap.h:
```diff
@@ -0,0 +1,950 @@
+	 * @param buf Optional response headers buffer
+	 */
+	void (*set_path)(struct bt_pbap_pce *pbap_pce, uint8_t rsp_code, struct net_buf *buf);
+
```
Done

lylezhu2012 commented on include/zephyr/bluetooth/classic/pbap.h:
```diff
@@ -0,0 +1,960 @@
+ *
+ *  @return 0 in case of success or negative value in case of error.
+ */
+int bt_pbap_pce_pull_vcardlisting(struct bt_pbap_pce *pbap_pce, struct net_buf *buf);
```
```suggestion
int bt_pbap_pce_pull_vcard_listing(struct bt_pbap_pce *pbap_pce, struct net_buf *buf);
```

lylezhu2012 commented on include/zephyr/bluetooth/classic/pbap.h:
```diff
@@ -0,0 +1,960 @@
+	 * @param pbap_pse The PBAP PSE object
+	 * @param buf      Request headers buffer (contains name, type, app parameters, etc.)
+	 */
+	void (*pull_vcardlisting)(struct bt_pbap_pse *pbap_pse, struct net_buf *buf);
```
```suggestion
	void (*pull_vcard_listing)(struct bt_pbap_pse *pbap_pse, struct net_buf *buf);
```

lylezhu2012 commented on include/zephyr/bluetooth/classic/pbap.h:
```diff
@@ -0,0 +1,960 @@
+ */
+int bt_pbap_verify_authentication(uint8_t nonce[BT_OBEX_CHALLENGE_TAG_NONCE_LEN],
+			uint8_t rsp_digest[BT_OBEX_RESPONSE_TAG_REQ_DIGEST_LEN],
+			const uint8_t *pwd);
```
```suggestion
				  uint8_t rsp_digest[BT_OBEX_RESPONSE_TAG_REQ_DIGEST_LEN],
				  const uint8_t *pwd);
```

lylezhu2012 commented on include/zephyr/bluetooth/classic/pbap.h:
```diff
@@ -0,0 +1,960 @@
+	 * @param rsp_code Response code @ref bt_pbap_rsp_code
+	 * @param buf Optional response headers buffer
+	 */
+	void (*set_phone_book)(struct bt_pbap_pce *pbap_pce, uint8_t rsp_code, struct net_buf *buf);
```
```suggestion
	void (*set_phonebook)(struct bt_pbap_pce *pbap_pce, uint8_t rsp_code, struct net_buf *buf);
```

lylezhu2012 commented on include/zephyr/bluetooth/classic/pbap.h:
```diff
@@ -0,0 +1,960 @@
+
+/** @brief MIME type for pull vCard entry operation */
+#define BT_PBAP_PULL_VCARD_ENTRY_TYPE "x-bt/vcard"
+
```
```suggestion
/** @brief OBEX Type header for pull phonebook operation */
#define BT_PBAP_PULL_PHONEBOOK_TYPE "x-bt/phonebook"

/** @brief OBEX Type header for pull vCard listing operation */
#define BT_PBAP_PULL_VCARD_LISTING_TYPE "x-bt/vcard-listing"

/** @brief OBEX Type header for pull vCard entry operation */
#define BT_PBAP_PULL_VCARD_ENTRY_TYPE "x-bt/vcard"

```

lylezhu2012 commented on subsys/bluetooth/host/classic/pbap.c:
```diff
@@ -0,0 +1,1460 @@
+	}
+
+	atomic_set(&pbap_pce->_state, BT_PBAP_STATE_DISCONNECTING);
+	return 0;
```
```suggestion
	if (bt_obex_has_header(buf, BT_OBEX_HEADER_ID_CONN_ID)) {
		err = pbap_check_header_conn_id(pbap_pce->_conn_id, buf);
		if (err != 0) {
			LOG_ERR("Failed to check conn id header");
			goto failed;
		}
	} else {
		err = bt_obex_add_header_conn_id(buf, pbap_pce->_conn_id);
		if (err != 0) {
			LOG_ERR("Failed to add conn id header");
			goto failed;
		}
	}

	err = bt_obex_disconnect(&pbap_pce->_client, buf);
	if (err != 0) {
		LOG_ERR("Failed to send conn rsp %d", err);
		goto failed;
	}

	atomic_set(&pbap_pce->_state, BT_PBAP_STATE_DISCONNECTING);
	return 0;

failed:
	if (allocated) {
		net_buf_unref(buf);
	}
	return err;
```

lylezhu2012 commented on include/zephyr/bluetooth/classic/pbap.h:
```diff
@@ -0,0 +1,1007 @@
+	const char *_optype;
+
+	/** @internal Flags. */
+	atomic_t _flags;
```
```suggestion
	/** @brief Callback operations structure. */
	struct bt_pbap_pse_cb *cb;

	/** @internal GOEP (Generic Object Exchange Profile) instance. */
	struct bt_goep _goep;

	/** @internal Transport state (atomic). */
	atomic_t _transport_state;

	/** @internal Connection ID.
	 *
	 *  Unique identifier for the OBEX connection session with the PCE.
	 */
	uint32_t _conn_id;

	/** @internal OBEX server handle.
	 *
	 *  Internal OBEX server structure for managing OBEX protocol operations.
	 */
	struct bt_obex_server _server;

	/** @internal PBAP connection state.
	 *
	 *  Atomic variable tracking the current state of the PBAP connection.
	 */
	atomic_t _state;

	/** @internal Pending request callback. */
	void (*_req_cb)(struct bt_pbap_pse *pse, struct net_buf *buf);

	/** @internal Current operation type string. */
	const char *_optype;

	/** @internal Flags. */
	atomic_t _flags;
```

lylezhu2012 commented on include/zephyr/bluetooth/classic/pbap.h:
```diff
@@ -0,0 +1,1007 @@
+ *
+ *  @param pwd Password string used for authentication (null-terminated).
+ *  @param nonce Output buffer to store the generated nonce (@ref BT_OBEX_CHALLENGE_TAG_NONCE_LEN
+ * bytes).
```
```suggestion
 *  @param nonce Output buffer to store the generated nonce (@ref BT_OBEX_CHALLENGE_TAG_NONCE_LEN
 *               bytes).
```

lylezhu2012 commented on subsys/bluetooth/host/classic/pbap.c:
```diff
@@ -0,0 +1,1468 @@
+
+	pbap_pce = CONTAINER_OF(client, struct bt_pbap_pce, _client);
+
+	if (pbap_pce->cb != NULL && pbap_pce->cb->abort) {
```
```suggestion
	if (pbap_pce->cb != NULL && pbap_pce->cb->abort != NULL) {
```

lylezhu2012 commented on subsys/bluetooth/host/classic/pbap_internal.h:
```diff
@@ -0,0 +1,35 @@
+	/** PBAP ready for upper layer traffic on it */
+	BT_PBAP_STATE_CONNECTED = 2,
+	/** PBAP in connecting state */
+	BT_PBAP_STATE_CONNECTING = 3,
```
```suggestion
	/** PBAP disconnected */
	BT_PBAP_STATE_DISCONNECTED = 0,
	/** PBAP in connecting state */
	BT_PBAP_STATE_CONNECTING = 1,
	/** PBAP ready for upper layer traffic on it */
	BT_PBAP_STATE_CONNECTED = 2,
	/** PBAP disconnecting */
	BT_PBAP_STATE_DISCONNECTING = 3,
```

lylezhu2012 commented on subsys/bluetooth/host/classic/shell/pbap.c:
```diff
@@ -0,0 +1,2665 @@
+{
+	uint8_t src_len;
+	uint8_t required_size;
+	uint8_t i;
```
The potential issue is that the local variables may overflow.

```suggestion
static size_t pbap_pce_ascii_to_unicode(const char *src, uint8_t *dest, size_t dest_size,
					 bool big_endian)
{
	size_t src_len;
	size_t required_size;
	size_t i;
```

lylezhu2012 commented on subsys/bluetooth/host/classic/shell/pbap.c:
```diff
@@ -0,0 +1,2665 @@
+					 uint16_t dest_size, bool big_endian)
+{
+	uint8_t ascii_len;
+	uint8_t i;
```
Similar issue here.

```suggestion
static size_t pbap_pse_unicode_to_ascii(const uint8_t *src, uint16_t src_len, char *dest,
					 uint16_t dest_size, bool big_endian)
{
	size_t ascii_len;
	size_t i;
```

lylezhu2012 commented on subsys/bluetooth/host/classic/shell/pbap.c:
```diff
@@ -0,0 +1,2665 @@
+	*value = total;
+	return 0;
+}
+
```
There is function for this purpose.

https://github.com/zephyrproject-rtos/zephyr/blob/7bc9e93d3f67986fe0f9989bbd8229f3d21a9be1/include/zephyr/shell/shell_string_conv.h#L51-L66

lylezhu2012 commented on subsys/bluetooth/host/classic/shell/pbap.c:
```diff
@@ -0,0 +1,2665 @@
+
+static struct bt_pbap_app *g_pbap_allocate(struct bt_conn *conn)
+{
+	for (uint8_t i = 0; i < APPL_PBAP_MAX_COUNT; i++) {
```
```suggestion
	ARRAY_FOR_EACH(g_pbap, i) {
```

lylezhu2012 commented on subsys/bluetooth/host/classic/shell/pbap.c:
```diff
@@ -0,0 +1,2665 @@
+static uint8_t tlv_buffers[TLV_COUNT][TLV_BUFFER_SIZE];
+static uint8_t tlv_count;
+
+static void pbap_parse_appl_param(struct net_buf *buf);
```
If the function declaration can be removed?

lylezhu2012 commented on include/zephyr/bluetooth/classic/pbap.h:
```diff
@@ -0,0 +1,1250 @@
+#define BT_PBAP_PSE_SUPPORTED_REPOSITORIES                                                         \
+	(BT_PBAP_PSE_REPOSITORY_LOCAL_PHONE_BOOK_ENABLE | BT_PBAP_PSE_REPOSITORY_SIM_ENABLE |      \
+	 BT_PBAP_PSE_REPOSITORY_SPEED_DIAL_ENABLE | BT_PBAP_PSE_REPOSITORY_FAVORITES_ENABLE)
+
```
Is it possible to move this code block to internal header?

lylezhu2012 commented on subsys/bluetooth/host/classic/shell/pbap.c:
```diff
@@ -0,0 +1,2521 @@
+SHELL_STATIC_SUBCMD_SET_CREATE(
+	add_appl_param,
+	SHELL_CMD_ARG(Order, NULL, "<indexed:0x00,alphanumeric:0x01,phonetic:0x02> ", cmd_add_order,
+		      2, 0),
+	SHELL_CMD_ARG(SearchValue, NULL, "<text> ", cmd_add_search_value, 2, 0),
+	SHELL_CMD_ARG(SearchAttribute, NULL, "<Name:0x00,Number:0x01,Sound:0x02> ",
+		      cmd_add_search_property, 2, 0),
+	SHELL_CMD_ARG(MaxListCount, NULL, "<0x0000-0xFFFF> ", cmd_add_max_list_count, 2, 0),
+	SHELL_CMD_ARG(ListStartOffset, NULL, "<0x0000-0xFFFF> ", cmd_add_list_start_offset, 2, 0),
+	SHELL_CMD_ARG(PropertySelector, NULL, "<64 bits mask : bt_pbap_appl_param_property_mask>",
+		      cmd_add_property_selector, 2, 0),
+	SHELL_CMD_ARG(Format, NULL, "<V2.1:0x00,V3.0:0x01> ", cmd_add_format, 2, 0),
+	SHELL_CMD_ARG(PhonebookSize, NULL, "<0x0000-0xFFFF> ", cmd_add_phonebook_size, 2, 0),
+	SHELL_CMD_ARG(NewMissedCalls, NULL, "<0x00-0xFF> ", cmd_add_new_missed_calls, 2, 0),
+	SHELL_CMD_ARG(PrimaryFolderVersion, NULL, "<16bytes> ", cmd_add_primary_folder_version, 2,
+		      0),
+	SHELL_CMD_ARG(SecondaryFolderVersion, NULL, "<16bytes> ", cmd_add_secondary_folder_version,
+		      2, 0),
+	SHELL_CMD_ARG(vCardSelector, NULL, "<64 bits mask : bt_pbap_appl_param_property_mask>",
+		      cmd_add_vcard_selector, 2, 0),
+	SHELL_CMD_ARG(DatabaseIdentifier, NULL, "<16bytes> ", cmd_add_database_identifier, 2, 0),
+	SHELL_CMD_ARG(vCardSelectorOperator, NULL, "<OR:0x00,AND:0x01> ",
+		      cmd_add_vcard_selector_operator, 2, 0),
+	SHELL_CMD_ARG(ResetNewMissedCalls, NULL, "<Reset:0x01> ", cmd_add_reset_new_missed_calls, 2,
+		      0),
+	SHELL_CMD_ARG(PbapSupportedFeatures, NULL, "", cmd_add_pbap_supported_features, 1, 0),
+	SHELL_SUBCMD_SET_END);
+
```
```suggestion
	add_appl_param,
	SHELL_CMD_ARG(Order, NULL, "<indexed/alphanumeric/phonetic>", cmd_add_order,
		      2, 0),
	SHELL_CMD_ARG(SearchValue, NULL, "<text>", cmd_add_search_value, 2, 0),
	SHELL_CMD_ARG(SearchAttribute, NULL, "<name/number/sound>",
		      cmd_add_search_property, 2, 0),
	SHELL_CMD_ARG(MaxListCount, NULL, "<0x0000-0xffff>", cmd_add_max_list_count, 2, 0),
	SHELL_CMD_ARG(ListStartOffset, NULL, "<0x0000-0xffff>", cmd_add_list_start_offset, 2, 0),
	SHELL_CMD_ARG(PropertySelector, NULL, "<64 bits mask: bt_pbap_appl_param_property_mask>",
		      cmd_add_property_selector, 2, 0),
	SHELL_CMD_ARG(Format, NULL, "<v2.1/v3.0>", cmd_add_format, 2, 0),
	SHELL_CMD_ARG(PhonebookSize, NULL, "<0x0000-0xffff>", cmd_add_phonebook_size, 2, 0),
	SHELL_CMD_ARG(NewMissedCalls, NULL, "<0x00-0xff>", cmd_add_new_missed_calls, 2, 0),
	SHELL_CMD_ARG(PrimaryFolderVersion, NULL, "<16bytes>", cmd_add_primary_folder_version, 2,
		      0),
	SHELL_CMD_ARG(SecondaryFolderVersion, NULL, "<16bytes>", cmd_add_secondary_folder_version,
		      2, 0),
	SHELL_CMD_ARG(vCardSelector, NULL, "<64 bits mask: bt_pbap_appl_param_property_mask>",
		      cmd_add_vcard_selector, 2, 0),
	SHELL_CMD_ARG(DatabaseIdentifier, NULL, "<16bytes>", cmd_add_database_identifier, 2, 0),
	SHELL_CMD_ARG(vCardSelectorOperator, NULL, "<or/and>",
		      cmd_add_vcard_selector_operator, 2, 0),
	SHELL_CMD_ARG(ResetNewMissedCalls, NULL, "", cmd_add_reset_new_missed_calls, 1,
		      0),
	SHELL_CMD_ARG(PbapSupportedFeatures, NULL, "<version>", cmd_add_pbap_supported_features, 2, 0),
	SHELL_SUBCMD_SET_END);

```

lylezhu2012 commented on subsys/bluetooth/host/classic/shell/pbap.c:
```diff
@@ -0,0 +1,2521 @@
+	err = bt_pbap_pse_connect_rsp(&g_pbap_app->pbap_pse, mopl, rsp_code, g_pbap_app->tx_buf);
+	if (err != 0) {
+		shell_error(sh, "Fail to send conn rsp %d", err);
+		net_buf_unref(g_pbap_app->tx_buf);
```
`g_pbap_app->tx_buf` should be checked before using it here.

CC0918 replied:
Done

lylezhu2012 commented on subsys/bluetooth/host/classic/shell/pbap.c:
```diff
@@ -0,0 +1,2521 @@
+	err = bt_pbap_pse_disconnect_rsp(&g_pbap_app->pbap_pse, rsp_code, g_pbap_app->tx_buf);
+	if (err != 0) {
+		shell_error(sh, "Fail to send conn rsp %d", err);
+		net_buf_unref(g_pbap_app->tx_buf);
```
Ditto.

CC0918 replied:
Done

lylezhu2012 commented on subsys/bluetooth/host/classic/shell/pbap.c:
```diff
@@ -0,0 +1,2613 @@
+		      0),
+	SHELL_CMD_ARG(ResetNewMissedCalls, NULL, "", cmd_add_reset_new_missed_calls, 1, 0),
+	SHELL_CMD_ARG(PbapSupportedFeatures, NULL, "uint32_t", cmd_add_pbap_supported_features, 2,
+		      0),
```
I think the feature could be optional.  If it is not passed, the default value can be set.

```suggestion
	SHELL_CMD_ARG(PbapSupportedFeatures, NULL, "[supported features]", cmd_add_pbap_supported_features, 1,
		      1),
```

CC0918 replied:
Done

lylezhu2012 commented on doc/connectivity/bluetooth/shell/classic/pbap.rst:
```diff
@@ -0,0 +1,466 @@
+     DatabaseIdentifier               : <16bytes>
+     vCardSelectorOperator            : <OR:0x00,AND:0x01>
+     ResetNewMissedCalls              : <Reset:0x01>
+     PbapSupportedFeatures            : [none]
```
The readme file should also be updated.

CC0918 replied:
Done

CC0918 committed:
Bluetooth: Classic: Add PBAP (Phone Book Access Profile) support

Add Phone Book Access Profile (PBAP) implementation for Bluetooth Classic,
supporting both Phone Book Client Equipment (PCE) and Phone Book Server
Equipment (PSE) roles.

PBAP enables remote access to phone book data stored on a device,
allowing clients to retrieve contact information, call history, and
other phonebook-related data.

Key features:
- PCE (client) and PSE (server) role support
- RFCOMM and L2CAP transport layer support
- Pull phone book, vCard listing, and vCard entry operations
- Set phone book navigation (up/down/root)
- Abort operation support
- OBEX authentication (challenge-response with MD5)
- Application parameters for filtering and formatting
- SDP service registration and discovery
- Single Response Mode (SRM) support for GOEP v2
- Configurable PSE features (download, browsing, database identifier,
  folder version counters, vCard selector, enhanced missed calls, etc.)
- Configurable PSE repositories (local phonebook, SIM, speed dial,
  favorites)

The implementation includes:
- Core PBAP protocol handling (subsys/bluetooth/host/classic/pbap.c)
- Public API header (include/zephyr/bluetooth/classic/pbap.h)
- Internal definitions (subsys/bluetooth/host/classic/pbap_internal.h)
- Shell commands for testing and demonstration
  (subsys/bluetooth/host/classic/shell/pbap.c)
- Comprehensive shell documentation with usage examples
  (doc/connectivity/bluetooth/shell/classic/pbap.rst)
- Kconfig options for feature and repository configuration
- Build system integration (CMakeLists.txt)
- Logging support

Authentication uses PSA Crypto API (MD5) for nonce generation and
response digest calculation.

Signed-off-by: Cheng Chang <cheng.chang@nxp.com>

makeshi commented on subsys/bluetooth/host/classic/shell/pbap.c:
```diff
@@ -0,0 +1,2635 @@
+	err = bt_obex_add_header_srm_param(pbap_app->tx_buf, 0x01);
+	if (err != 0) {
+		bt_shell_error("Fail to add SRMP header %d", err);
+		return err;
```
For this failure case, does need net_buf_unref pbap_app->tx_buf?

CC0918 replied:
Do net_buf_unref pbap_app->tx_buf in pbap_pce_pull() and pbap_pull_rsp() which call it.
